Add support for where and join_where in many_to_many and has_many associations

jadlr opened the pull request:
We're using dataloader with a schema that has `where` and `join_where` clauses in many-to-many relationships.

Before this, these clauses where ignored.

jadlr commented:
Maybe you could also have a look, @mbuhot since you wrote the original code. Any input would be highly appreciated. 

I failed to create a test case that runs the functions in question.

mbuhot commented on lib/dataloader/ecto.ex:
```diff
@@ -786,6 +794,7 @@ if Code.ensure_loaded?(Ecto) do
         )
       end
 
+      # Fix me: never called because assoc is always length == 1
```
I think you would hit this case with a `has_many, through: [:some_many_to_many_assoc]`. Eg in `test/support/user.ex` file, something like:


```elixir
    many_to_many(:liked_posts, Dataloader.Post, join_through: Dataloader.Like)
    has_many(:liked_posts_scores, through: [:liked_posts, :scores])
```

jadlr replied:
Thanks a lot for the hint! I added a test for this in 3b17bac

I'll now adapt the `where` and `join_where` code to work with it.

jadlr replied:
To clarify: the added test only tests this case:

```elixir
      defp build_preload_lateral_query(
             [%Ecto.Association.ManyToMany{} = assoc],
             query,
             :join_last
           ) do
```

I'll see if I can come up with cases for the remaining untested cases.

mbuhot commented on lib/dataloader/ecto.ex:
```diff
@@ -752,13 +752,21 @@ if Code.ensure_loaded?(Ecto) do
+        binds = Ecto.Query.Builder.count_binds(join)
+
+        join
+        |> Ecto.Association.combine_joins_query(assoc.where, 0)
```
Probably worth posting to the ecto mailing list (or just send a PR) to see if we can get `Ecto.Association` to be a documented API.

https://github.com/elixir-ecto/ecto/blob/v3.5.8/lib/ecto/association.ex#L30-L31

benwilson512 commented:
@mbuhot thank you for the review!

jadlr commented:
@benwilson512 @mbuhot Thanks for the feedback!

We added tests for all of the lateral join builder functions and implemented `where` and `join_where` on all of them.

Please let us know if there is anything else we need to get this into main.

mbuhot commented:
Could we add a test for `has_many: .. where: ..` ?

Eg in `test/support/user.ex`

```elixir
has_many(:published_posts, Dataloader.Post, where: [status: "published"])
```

I suspect the remaining clauses of `build_preload_lateral_query` may also need to account for the `assoc.where` condition.

jadlr commented:
You're right, `where` should also work in `has_many` assocs. We didn't need that case, that's why we initially didn't implement it. 

I added the implementation and tests.

jadlr commented:
@benwilson512 @mbuhot could you please have another look? We'd love to get this back into main.

benwilson512 reviewed (approved):
The tests look good to me, the implementation involves more knowledge of ecto internals than I can weigh in on unfortunately without more time.

mbuhot commented:
@jadlr Looking good. The one thing that hasn't been covered is a conditional `belongs_to:` association.
It took me a while to confirm that it is indeed supported in ecto, as the [BelongsTo](https://hexdocs.pm/ecto/Ecto.Association.BelongsTo.html) struct docs don't mention the `where` field. 
But the [belongs_to/3](https://hexdocs.pm/ecto/Ecto.Schema.html#belongs_to/3) docs do mention it:

```
:where - A filter for the association. See "Filtering associations" in has_many/3.
```

The good news it I _think_ we can support it by deleting some code :) 

https://github.com/absinthe-graphql/dataloader/blob/a78ea35868b442a2157a3a5b73c53ac78f5a79d1/lib/dataloader/ecto.ex#L790-L799

Can be collapsed back into a single clause to cover `Has` and `BelongsTo` associations:

```elixir
      defp build_preload_lateral_query([assoc], query, :join_first) do
        query
        |> where([x], field(x, ^assoc.related_key) == field(parent_as(:parent), ^assoc.owner_key))
        |> Ecto.Association.combine_assoc_query(assoc.where)
      end
```

and similarly for https://github.com/absinthe-graphql/dataloader/blob/a78ea35868b442a2157a3a5b73c53ac78f5a79d1/lib/dataloader/ecto.ex#L801-L821

https://github.com/absinthe-graphql/dataloader/blob/a78ea35868b442a2157a3a5b73c53ac78f5a79d1/lib/dataloader/ecto.ex#L875-L898

https://github.com/absinthe-graphql/dataloader/blob/a78ea35868b442a2157a3a5b73c53ac78f5a79d1/lib/dataloader/ecto.ex#L900-L925

jadlr commented:
Thank you so much for the review, Mike!

I changed the code an added a test.

mbuhot reviewed (approved):
👍 👍 I think this is good to go @benwilson512

benwilson512 commented:
Thank you all!
